import { Tabs, Tab, Callout } from 'nextra/components';

## Install

<Tabs items={['yarn', 'pnpm', 'npm']} defaultIndex={0}>
    <Tab>
        ```bash
        yarn add @httpx/exception
        ```
    </Tab>
    <Tab>
        ```bash
        pnpm add @httpx/exception
        ```
    </Tab>
    <Tab>
        ```bash
        npm install @httpx/exception
        ```
    </Tab>
</Tabs>

<Callout type="info" emoji="ℹ️">
    You might want to install a polyfill for [Error.cause](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause)
    if you need to [support older browsers](https://caniuse.com/mdn-javascript_builtins_error_error_options_cause_parameter).
    ([error-cause-polyfill](https://github.com/ehmicky/error-cause-polyfill)). This is not needed on node >=16.9.0.
</Callout>


## Usage

### By named imports

Explicit named imports are prefixed by `Http` to ease
IDE experience. Message is optional and default to the [default message](#about-default-message). Additional
parameters are supported such as `url`, `method`, `code`, `errorId` and `cause`.

<Tabs items={['Basic', 'Extended params', 'With issues']} defaultIndex={0}>
    <Tab>
        ```typescript
        import { HttpNotFound, HttpBadRequest } from '@httpx/exception';

        const e1 = new HttpNotFound();
        e1.message;    // 'Not found'
        e2.statusCode; // 404

        const e2 = new HttpNotFound('Page not found');

        e2.message;    // 'Page not found'
        e2.statusCode; // 404

        const e400 = new HttpBadRequest('Problems parsing JSON');
        ```
    </Tab>
    <Tab>
        ```typescript
        import { HttpGatewayTimeout, HttpInternalServerError } from '@httpx/exception';

        const e500 = new HttpInternalServerError({
             url: 'https://api.dev/gateway',
             method: 'GET',
             code: 'ERR_UNREACHABLE_SERVICE',
             errorId: nanoid(),
             cause: new HttpGatewayTimeout({
                message: 'This Serverless Function has timed out',
                errorId: 'cdg1::h99k2-1664884491087-b41a2832f559',
             }),
        });

        e500.message;     // 'Internal server error'
        e500.statusCode;  // 500
        e500.url;         // 'https://api.dev/gateway'
        e500.method;      // 'GET'
        // ...
        ```

    </Tab>
    <Tab>

        ```typescript
        import { HttpUnprocessableEntity } from '@httpx/exception';

        const e422 = new HttpUnprocessableEntity({
           message: 'Request validation failed',
           issues: [   // typed as ValidationIssues[]
                {
                    message: 'Invalid email',
                    path: 'email',
                    code: 'invalid_email',
                },
                {
                    message: 'Invalid address',
                    path: ['addresses', 0, 'line1'],
                    code: 'empty_string',
                },
          ]
        });
        ```

        <Callout type="info" emoji="ℹ️">
            According to MDN, the 422-HttpUnprocessableEntity status code is reserved for WebDAV, but it is widely used in REST APIs
            (ie: [github](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#client-errors]), rails...).

            The choice to add validation errors to the 422 exception might be controversial and is optional. However, it offers
            some benefits as a global error handler can leverage them to generate error responses
            (ie: [json-api error responses](https://jsonapi.org/examples/#error-objects-basics)).
        </Callout>
    </Tab>

</Tabs>

### By status code

The `createHttpException` function allows to create an exception from an
arbitrary status code.

```typescript
import { createHttpException } from '@httpx/exception';

const e404 = createHttpException(404); // e404 instanceof HttpClientException
const e500 = createHttpException(500); // e500 instanceof HttpServerException
```

Additional [parameters](#httpexception-parameters) can be provided as a second argument.

```typescript
throw createHttpException(404, 'The graal is yet to find !');

throw createHttpException(500, {
  message: 'Something really wrong happened.',
  url: 'https://api.dev/gateway',
  cause: new HttpNotImplemented(), // or any Error...
});
```

<Callout type="warning">
    No checks are done about the validity of the provided status code. See also
    [about non-official status codes](#non-official-status-codes)
</Callout>


```typescript
import { HttpNotFound } from '@httpx/exception';

// Simple
throw new HttpNotFound(); // message = 'Not found'

// Custom message
throw new HttpNotFound('User not found');

import {
  HttpException,
  HttpGatewayTimeout,
  HttpInternalServerError,
} from '@httpx/exception';

// Another way (could be with ts-result, either...)
const fetchData = async <T,>(url: string): Promise<T | HttpException> => {
  return new HttpInternalServerError({
    url,
    cause: new HttpGatewayTimeout({
      code: 'This Serverless Function has timed out',
      errorId: 'cdg1::h99k2-1664884491087-b41a2832f559',
    }),
  });
};

const data = await fetchData<{ data: unknown }>('https://api.dev/gateway');
if (data instanceof Error) {
  // handle error
} else {
}
```

#### HttpException parameters

Http exception optionally accepts a parameter of type `string | HttpExceptionParams`. If no parameter
is provided a [default message](#about-default-message) will be set.
When a `string` is provided it will be used as the error message, otherwise you can use the following params:

| HttpExceptionParams | Type      | Description                                                                                                                                                           |
| ------------------- | --------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| message             | `string?` | [Error.message](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/message), see [about default message](#about-default-message). |
| url                 | `string?` | Origin url ([about context](#about-context)).                                                                                                                         |
| method              | `string?` | Origin http method ([about context](#about-context)).                                                                                                                 |
| code                | `string?` | Custom code ([about context](#about-context)).                                                                                                                        |
| errorId             | `string?` | Unique id ([about context](#about-context)).                                                                                                                          |
| cause               | `Error?`  | Error.cause, see also [about error cause](#about-errorcause).                                                                                                         |

Example:

#### HttpException properties

| HttpException | Type                 | Description                                                                                                                        |
| ------------- | -------------------- | ---------------------------------------------------------------------------------------------------------------------------------- |
| statusCode    | `number`             | Http error status code (400-599).                                                                                                  |
| message       | `string`             | Default or provided message.                                                                                                       |
| url           | `string?`            | Origin url ([about context](#about-context)).                                                                                      |
| method        | `string?`            | Origin http method ([about context](#about-context)).                                                                              |
| code          | `string?`            | Custom code ([about context](#about-context)).                                                                                     |
| errorId       | `string?`            | Unique id ([about context](#about-context)).                                                                                       |
| stack         | `string?`            | @see [Error.prototype.stack](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Stack) on MDN. |
| cause         | `Error?`             | @see [about error cause](#about-errorcause)                                                                                        |
| issues?       | `ValidationIssues[]` | Only supported by HttpUnprocessableEntity (422)                                                                                    |
